var/spack/repos/builtin/packages/tfel/package.py

```diff
@@ -127,6 +131,11 @@ def cmake_args(self):
         else:
             args.append("-Denable-python-bindings=OFF")
 
+        if '+portable_build' in self.spec:
+            args.append("-Denable-portable-build=ON")
+        else:
+            args.append("-Denable-portable-build=OFF")
+
```

Comment:
I think `-march=native` should be always disabled, since our compiler wrappers inject thise flags automatically based on the target of the spec.

Reply:
Well, this is no done by default in `TFEL` nor `MGIS` because most users install from sources but don't want to bother with defining the `CXXFLAGS`.
However, we have a special flag that requires `cmake` to use to the `CXXFLAGS`  environment variables (for the `conda` package). If you want, I can add a variant related to this flag and enable this by default.
I would appreciate if we could validate the PR as it is and make this change in a few days (See this ticket https://sourceforge.net/p/tfel/tickets/257/).

Reply:
> Well, this is no done by default in TFEL nor MGIS because most users install from sources but don't want to bother with defining the CXXFLAGS

I think there's a misunderstanding. I don't mean to interfere with upstream choices, I just wanted to say that in my opinion the `portable_build` variant should be removed from `package.py` and we should always use:
```python
args.append("-Denable-portable-build=ON")
```
When compiling a code with Spack, you're implicitly using Spack's compiler wrappers that *already* inject microarchitecture specific flags based on the `target` attribute of the spec.

Reply:
@alalazo Well, there was request from my users to add this option because they wanted to use `spack` to install `TFEL` on a sever based on heterogeneous architectures. But by default, it' enabled !

Reply:
@thelfer I don't get your point, sorry. I'll rephrase. Try not to add any flag that specify how to emit code for microarchitectures, because Spack takes care of it and the recipe may be interfering. To be more clear, this is what Spack does when compiling zlib for broadwell:
```console
[cc] /usr/bin/gcc-10 -march=broadwell -mtune=broadwell -c ztest17714.c
[cc] /usr/bin/gcc-10 -march=broadwell -mtune=broadwell -c -fPIC -O2 ztest17714.c
[ ... ]
```
If users want a portable installation they just need to say:
```console
spack install zlib target=x86_64
```
This holds true only if the package does not try to use `-march` type flags.

Reply:
@alalazo Sorry I was confused. What about other flags (`-02`, etc...) ?

Reply:
@thelfer Those are fine since they don't influence which assembly instructions will be emitted.

Reply:
@alalazo Ok. Concerning compiler flags, my packages have an all or nothing approach. i.e.:

- by default, CXXFLAGS is discarded and `march=native` is set
- with -Denable-portable-build=ON, CXXFLAGS is discarded but `march=native` is not set.
- with another option, we use CXXFLAGS but do not add anything to it.

Reply:
> with -Denable-portable-build=ON, CXXFLAGS is discarded but march=native is not set.

If your `CMakelists.txt` respects the default settings for [build_type](https://cmake.org/cmake/help/v3.19/variable/CMAKE_BUILD_TYPE.html) I think that may be sufficient, but of course you have much more domain knowledge on tfel requirements than I do :slightly_smiling_face:

Reply:
@alalazo Yes, we follow the default settings for `build_type` !

Reply:
@alalazo So, anything missing for the PR to be accepted ?

Reply:
I was waiting for the removal of the `portable_build` variant in the package, and unconditional use of:
```python
args.append("-Denable-portable-build=ON")
```
since, if I understand correctly, `CMAKE_BUILD_TYPE=Release` etc. will be honored. Let me know if I missed some detail in the discussion.

Reply:
@alalazo I would like, but if I remove this, my users won't be happy because they use spack to install `TFEL` on heterogenous servers...

Reply:
Spack should be able to take care of that. The variant is redundant. If you remove the variant and do:
```console
$ spack -d install tfel target=x86_64
```
what you should see in the debugging logs that get generated is that the software is compiled with:
```
--march=x86-64 -mtune=generic
```

Reply:
@alalazo I changed my packages to enable by default the use of external compiler flags. That shall do the trick. I have made various tests outside spack and it works as expected. To be sure, how are passed the march flags to `cmake` ?

Reply:
The mechanism doesn't depend on CMake. Spack tells CMake to use its compiler wrappers the wrapper layer adds the flags transparently.